src/hir/combiner.rs

```diff
@@ -157,3 +170,162 @@ impl Combiner {
         Ok(Hir::Root(root))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use anyhow::{Error, Result};
+    use pretty_assertions::assert_eq;
+    use std::panic::catch_unwind;
+
+    use crate::hir::{self, Hir};
+    use crate::scaffold::modifiers;
+    use crate::span::{Position, Span};
+    use crate::syntax::parser::Parser;
+    use crate::syntax::tokenizer::Tokenizer;
+
+    fn translate(text: &str) -> Result<Hir> {
+        let tokens = Tokenizer::new().tokenize(&text)?;
+        let ast = Parser::new().parse(&text, &tokens)?;
+        let mut discoverer = modifiers::ModifierDiscoverer::new();
+        let modifiers = discoverer.discover(&ast);
+
+        Ok(hir::translator::Translator::new().translate(&ast, modifiers))
+    }
+
+    fn combine(hirs: &Vec<Hir>) -> Result<Hir, Error> {
+        Ok(crate::hir::combiner::Combiner::new().combine(&hirs)?)
+    }
+
+    fn root(children: Vec<Hir>) -> Hir {
+        Hir::Root(hir::Root { children })
+    }
+
+    fn contract(identifier: String, children: Vec<Hir>) -> Hir {
+        Hir::ContractDefinition(hir::ContractDefinition {
+            identifier,
+            children,
+        })
+    }
+
+    fn function(
+        identifier: String,
+        ty: hir::FunctionTy,
+        span: Span,
+        modifiers: Option<Vec<String>>,
+        children: Option<Vec<Hir>>,
+    ) -> Hir {
+        Hir::FunctionDefinition(hir::FunctionDefinition {
+            identifier,
+            ty,
+            span,
+            modifiers,
+            children,
+        })
+    }
+
+    fn comment(lexeme: String) -> Hir {
+        Hir::Comment(hir::Comment { lexeme })
+    }
+
+    #[test]
+    fn panics_when_root_contract_identifier_is_missing() -> Result<()> {
+        let trees = vec![
+            "::orphanedFunction\n└── when something bad happens\n   └── it should revert",
+            "Contract::function\n└── when something bad happens\n   └── it should revert",
+        ];
+        let hirs = trees
+            .iter()
+            .map(|tree| translate(tree))
+            .collect::<Result<Vec<Hir>>>()?;
+
+        let result = catch_unwind(|| combine(&hirs));
+        assert!(result.is_err());
+        assert_eq!(
+            result
+                .unwrap_err()
+                .downcast_ref::<String>()
+                .unwrap()
+                .as_str(),
+            "expected contract identifier at tree root"
+        );
```

Comment:
Oh, this is neat. However, you can just annotate the test with `#[should_panic]`.
This is a tiny bit more precise so I'm ok with it 👍🏻

Reply:
Thanks for the heads up 🙏